lib/puppet/application/cert.rb

```diff
@@ -264,7 +264,9 @@ def main
       hosts = command_line.args.collect { |h| h.downcase }
     end
     begin
-      apply(@ca, :revoke, options.merge(:to => hosts)) if subcommand == :destroy
+      if subcommand == :destroy and (@ca.waiting? & hosts).empty?
+        apply(@ca, :revoke, options.merge(:to => hosts))
+      end
```

Comment:
If we clean multiple hosts, but at least one host only has a CSR, then we will fail to revoke the hosts that have issued certs:

```
$ bx puppet cert list --all
  "agent3"    (SHA256) 8C:04:18:2D:66:6D:06:64:2B:8A:CA:11:9A:6F:56:A1:D9:E3:2D:D9:C6:D0:F0:F1:49:32:57:99:6F:DB:37:50
  "agent4"    (SHA256) 65:CE:2F:F2:D2:C5:DE:8D:31:46:5B:1E:FF:7F:EA:8A:B3:37:3A:FD:89:1E:AB:B5:B5:60:BC:CF:3E:FE:9F:D2
  "agent5"    (SHA256) 6B:90:A3:00:FC:35:27:37:FF:38:6D:0A:B5:63:61:E8:1A:C3:F0:43:6C:BF:7C:8A:80:36:51:AD:1D:BF:5C:6D
+ "agent6"    (SHA256) 3A:F5:2F:1B:8E:FB:C4:67:4A:FB:A1:86:AD:4A:11:F7:20:A0:21:2F:54:57:CE:57:A9:8F:3C:DC:EA:57:67:88
+ "localhost" (SHA256) C7:11:7C:93:49:DE:58:FD:90:4F:AD:5F:FC:39:2B:9C:AE:05:FB:F7:86:49:D1:CB:4D:F6:22:B3:1B:BE:12:8F
$ bx puppet cert clean agent5 agent6
Notice: Removing file Puppet::SSL::CertificateRequest agent5 at '/Users/josh/.puppetlabs/etc/puppet/ssl/ca/requests/agent5.pem'
Notice: Removing file Puppet::SSL::CertificateRequest agent5 at '/Users/josh/.puppetlabs/etc/puppet/ssl/certificate_requests/agent5.pem'
Notice: Removing file Puppet::SSL::Key agent5 at '/Users/josh/.puppetlabs/etc/puppet/ssl/private_keys/agent5.pem'
Notice: Removing file Puppet::SSL::Certificate agent6 at '/Users/josh/.puppetlabs/etc/puppet/ssl/ca/signed/agent6.pem'
Notice: Removing file Puppet::SSL::CertificateRequest agent6 at '/Users/josh/.puppetlabs/etc/puppet/ssl/certificate_requests/agent6.pem'
Notice: Removing file Puppet::SSL::Key agent6 at '/Users/josh/.puppetlabs/etc/puppet/ssl/private_keys/agent6.pem'
$ openssl crl -in ~/.puppetlabs/etc/puppet/ssl/ca/ca_crl.pem -noout -text
Certificate Revocation List (CRL):
        Version 2 (0x1)
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: /CN=Puppet CA: localhost
        Last Update: Jan 26 19:55:59 2018 GMT
        Next Update: Jan 25 19:56:00 2023 GMT
        CRL extensions:
            X509v3 Authority Key Identifier:
                keyid:AB:31:1E:8E:32:14:B8:F0:EC:63:30:E7:66:74:EE:76:13:A9:78:FA

            X509v3 CRL Number:
                0
No Revoked Certificates.
...
```

I think we need to partition `hosts` into those that are `waiting?` and those that aren't, and call `clean` on the latter group.

Reply:
@joshcooper good catch, I pushed another commit to rectify the situation.  Not sure if it's the most elegant solution but it seems to work correctly now.